src/snowflake/snowpark/_internal/data_source_utils.py

```diff
@@ -0,0 +1,363 @@
+
+    return StructType(fields)
+
+
+def infer_data_source_schema(
+    conn: Connection, table: str, dbms_type: DBMS_TYPE, driver_info: str
+) -> StructType:
+    try:
+        cursor = conn.cursor()
+        if dbms_type == DBMS_TYPE.SQL_SERVER_DB:
+            query = f"""
+                    SELECT COLUMN_NAME, DATA_TYPE, NUMERIC_PRECISION, NUMERIC_SCALE, IS_NULLABLE
+                    FROM INFORMATION_SCHEMA.COLUMNS
+                    WHERE TABLE_NAME = '{table}'
+                    """
+            cursor.execute(query)
+            raw_schema = cursor.fetchall()
+            return sql_server_to_snowpark_type(raw_schema)
+        elif dbms_type == DBMS_TYPE.ORACLE_DB:
+            query = f"""
+                    SELECT COLUMN_NAME, DATA_TYPE, DATA_PRECISION, DATA_SCALE, NULLABLE
+                    FROM USER_TAB_COLUMNS
+                    WHERE TABLE_NAME = '{table}'
+                    """
+            cursor.execute(query)
+            raw_schema = cursor.fetchall()
+            return oracledb_to_snowpark_type(raw_schema)
+        else:
+            raise NotImplementedError(
+                f"Failed to infer Snowpark DataFrame schema from source '{table}'. "
+                f"Currently supported drivers are 'pyodbc' and 'oracledb', but got: '{driver_info}'. "
+                "To avoid auto inference, you can manually specify the Snowpark DataFrame schema using 'custom_schema' in DataFrameReader.dbapi."
+            )
+
+    except Exception as exc:
+        raise SnowparkDataframeReaderException(
+            f"Failed to infer Snowpark DataFrame schema from table '{table}'. To avoid auto inference, you can manually specify the Snowpark DataFrame schema using 'custom_schema' in DataFrameReader.dbapi."
+        ) from exc
+
+
+def data_source_data_to_pandas_df(
+    data: List[Any], schema: StructType, current_db: str, driver_info: str
+) -> pd.DataFrame:
+    columns = [col.name for col in schema.fields]
+    df = pd.DataFrame.from_records(data, columns=columns)
+
+    # convert timestamp and date to string to work around SNOW-1911989
+    df = df.map(
+        lambda x: x.isoformat()
+        if isinstance(x, (datetime.datetime, datetime.date))
+        else x
+    )
+    # convert binary type to object type to work around SNOW-1912094
+    df = df.map(lambda x: x.hex() if isinstance(x, (bytearray, bytes)) else x)
+    if current_db == DBMS_TYPE.SQL_SERVER_DB or current_db == DBMS_TYPE.SQLITE_DB:
+        return df
+    elif current_db == DBMS_TYPE.ORACLE_DB:
+        # apply read to LOB object, we currently have FakeOracleLOB because CLOB and BLOB is represented by an
+        # oracledb object and we cannot add it as our dependency in test, so we fake it in this way
+        # TODO: SNOW-1923698 remove FakeOracleLOB after we have test environment
+        df = df.map(lambda x: x.read() if (type(x).__name__.lower() == "lob") else x)
```

Comment:
we can delete from 306 to 312 since the oracle perf optimization is introduced, this piece of code is only meant to hanlde LOB object while it is handled elsewhere already

Reply:
we can actually delete until line 317